File: reconstruct-go/internal/config/load.go
package config

// 为什么存在：换 DeepSeek / 端口不应改 loop 或 Gin 路由；配置与 TypeScript 版同一套 JSON。
// 功能作用：读 reconstruct-go/config.json，再用 config.local.json 覆盖。若本目录没有密钥，会读 ../reconstruct/config.local.json。

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

type AppConfig struct {
	APIKey       string
	BaseURL      string
	Model        string
	SystemPrompt string
	Listen       string
}

type fileShape struct {
	BaseURL      string `json:"baseURL"`
	Model        string `json:"model"`
	SystemPrompt string `json:"systemPrompt"`
	APIKeyEnv    string `json:"apiKeyEnv"`
	APIKey       string `json:"apiKey"`
	Listen       string `json:"listen"`
}

func rootDir() string {
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		return "."
	}
	return filepath.Clean(filepath.Join(filepath.Dir(file), "../.."))
}

func readJSON(path string) (fileShape, error) {
	var out fileShape
	raw, err := os.ReadFile(path)
	if err != nil {
		return out, err
	}
	if err := json.Unmarshal(raw, &out); err != nil {
		return out, fmt.Errorf("%s: %w", path, err)
	}
	return out, nil
}

func pick(values ...string) string {
	for _, v := range values {
		if strings.TrimSpace(v) != "" {
			return strings.TrimSpace(v)
		}
	}
	return ""
}

func Load() (AppConfig, error) {
	root := rootDir()
	shared, err := readJSON(filepath.Join(root, "config.json"))
	if err != nil {
		return AppConfig{}, err
	}

	local := fileShape{}
	localPath := filepath.Join(root, "config.local.json")
	if _, err := os.Stat(localPath); err == nil {
		local, err = readJSON(localPath)
		if err != nil {
			return AppConfig{}, err
		}
	} else {
		sibling := filepath.Join(root, "..", "reconstruct", "config.local.json")
		if _, err := os.Stat(sibling); err == nil {
			local, err = readJSON(sibling)
			if err != nil {
				return AppConfig{}, err
			}
		}
	}

	baseURL := pick(local.BaseURL, shared.BaseURL)
	model := pick(local.Model, shared.Model)
	systemPrompt := pick(local.SystemPrompt, shared.SystemPrompt)
	apiKeyEnv := pick(local.APIKeyEnv, shared.APIKeyEnv)
	listen := pick(local.Listen, shared.Listen, "127.0.0.1:8080")
	if baseURL == "" || model == "" || systemPrompt == "" || apiKeyEnv == "" {
		return AppConfig{}, fmt.Errorf("config.json 需要 baseURL、model、systemPrompt、apiKeyEnv")
	}

	apiKey := pick(local.APIKey, os.Getenv(apiKeyEnv))
	if apiKey == "" {
		return AppConfig{}, fmt.Errorf("缺少 API 密钥。复制 config.local.example.json 为 config.local.json，或沿用 reconstruct/config.local.json，或设置 %s", apiKeyEnv)
	}

	return AppConfig{
		APIKey:       apiKey,
		BaseURL:      baseURL,
		Model:        model,
		SystemPrompt: systemPrompt,
		Listen:       listen,
	}, nil
}

File: reconstruct/src/config/load.ts
/**
 * 为什么存在：换 DeepSeek / 别的 OpenAI-compatible 端点不应改 loop；endpoint、模型、密钥来源放在 reconstruct 根目录的 JSON 里。
 * 功能作用：读 config.json，再用可选的 config.local.json 覆盖（可写 apiKey）。密钥不要提交。
 */
import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

export type AppConfig = {
	apiKey: string;
	baseURL: string;
	model: string;
	systemPrompt: string;
};

type FileShape = {
	baseURL?: unknown;
	model?: unknown;
	systemPrompt?: unknown;
	apiKeyEnv?: unknown;
	apiKey?: unknown;
};

/** 为什么存在：session 文件要固定落在 reconstruct/.sessions，不跟调用时的 cwd 乱跑。 */
export const reconstructRoot = join(dirname(fileURLToPath(import.meta.url)), "../..");

function readJson(path: string): FileShape {
	const raw: unknown = JSON.parse(readFileSync(path, "utf8"));
	if (typeof raw !== "object" || raw === null) {
		throw new Error(`${path} 根节点必须是对象`);
	}
	return raw as FileShape;
}

function asString(value: unknown, label: string): string {
	if (typeof value !== "string" || value.trim() === "") {
		throw new Error(`配置缺少有效字符串: ${label}`);
	}
	return value;
}

export function loadAppConfig(): AppConfig {
	const sharedPath = join(reconstructRoot, "config.json");
	const localPath = join(reconstructRoot, "config.local.json");
	const shared = readJson(sharedPath);
	const local = existsSync(localPath) ? readJson(localPath) : {};

	const baseURL = asString(local.baseURL ?? shared.baseURL, "baseURL");
	const model = asString(local.model ?? shared.model, "model");
	const systemPrompt = asString(local.systemPrompt ?? shared.systemPrompt, "systemPrompt");
	const apiKeyEnv = asString(local.apiKeyEnv ?? shared.apiKeyEnv, "apiKeyEnv");

	const apiKeyFromFile = typeof local.apiKey === "string" ? local.apiKey.trim() : "";
	const apiKey = apiKeyFromFile !== "" ? apiKeyFromFile : (process.env[apiKeyEnv] ?? "").trim();
	if (apiKey === "") {
		throw new Error(
			`缺少 API 密钥。复制 config.local.example.json 为 config.local.json 并填入 apiKey，或设置环境变量 ${apiKeyEnv}。`,
		);
	}

	return { apiKey, baseURL, model, systemPrompt };
}
